alf/algorithms/mbrl_algorithm.py

```diff
@@ -275,3 +280,132 @@ def calc_loss(self, experience, training_info: MbrlInfo):
+            reward_module=None,
+            planner_module=planner_module,
+            planner_optimizer=planner_optimizer,
+            env=env,
+            config=config,
+            debug_summaries=debug_summaries,
+            name=name)
+
+        flat_action_spec = nest.flatten(action_spec)
+        action_spec = flat_action_spec[0]
+
+        assert action_spec.is_continuous, "only support \
+                                                    continious control"
+
+        num_actions = action_spec.shape[-1]
+
+        self._action_spec = action_spec
+        self._num_actions = num_actions
+
+        self._latent_pred_rep_module = None  # set it later
+
+    def set_latent_predictive_representation_module(
+            self, latent_pred_rep_module: PredictiveRepresentationLearner):
+        self._latent_pred_rep_module = latent_pred_rep_module
+
+    def _trainable_attributes_to_ignore(self):
+        return ['_latent_pred_rep_module']
+
+    @torch.no_grad()
+    def _predict_multi_step_cost(self, init_rep, actions):
+        """Compute the total cost by unrolling multiple steps according to
+            the given initial observation and multi-step actions.
+        Args:
+            init_rep: the current observation for predicting quantities of
+                future time steps of shape [B, d]
+            actions (Tensor): a set of action sequences to
+                shape [B, population, unroll_steps, action_dim]
+        Returns:
+            cost (Tensor): negation of accumulated predicted reward, with
+                the shape of [B, population]
+        """
+        batch_size, population_size, num_unroll_steps = actions.shape[0:3]
+
+        init_rep = self._expand_to_population(init_rep, population_size)
+
+        # merge batch with population
+        # [B, population, unroll_steps, ...] -> [B*population, unroll_steps, ...]
+        actions = torch.reshape(actions, (-1, *actions.shape[2:]))
+
+        pred_rewards = self._latent_pred_rep_module.predict_multi_step(
+            init_rep, actions)
+
+        pred_rewards = pred_rewards.view(num_unroll_steps, batch_size,
+                                         population_size, -1)
+        # [B, population, unroll_steps, reward_dim]
+        pred_rewards = pred_rewards.permute(1, 2, 0, 3)
+
+        # currently assume the first dimension is the overall reward
+        # [B, population, unroll_steps]
+        pred_rewards = pred_rewards[..., 0]
+        cost = -pred_rewards
+        cost = cost.sum(2)
+        return cost
+
+    def _predict_with_planning(self, time_step: TimeStep, state,
+                               epsilon_greedy):
+        action, state = self._planner_module.predict_plan(
+            time_step, state, epsilon_greedy)
+
+        return AlgStep(output=action, state=state, info=MbrlInfo())
+
+    def train_step(self, exp: Experience, state: MbrlState):
+        return AlgStep(output=(), state=state, info=MbrlInfo())
```

Comment:
you can return `AlgStep(output=(), state=state, info=LossInfo())` and omit `calc_loss()`

Reply:
The dummy ``train_step`` and ``calc_cost`` are used to overwrite those functions from the base ``MbrlAlgorithm``. Comments added.